osgbuild/fetch_sources.py

```diff
@@ -32,194 +82,263 @@
+def _nvl(arg, default):
+    return default if arg is None else arg
 
-        orig_dir = os.getcwd()
-        os.chdir(checkout_dir)
-        try:
-            if tag:
-                output, rc = utils.sbacktick(["git", "show-ref", tag])
-                if rc:
-                    raise Error("Repository %s does not contain a tag named %s." % (git_url, tag))
-                sha1 = output.split()[0]
-                if sha1 != git_hash and deref_git_sha(sha1) != deref_git_sha(git_hash):
-                    msg = "Hash mismatch for %s tag %s\n    expected: %s\n    actual:   %s" % \
-                          (git_url, tag, git_hash, sha1)
-                    if nocheck:
-                        log.warning(msg + "\n    (ignored)")
-                        git_hash = sha1
-                    else:
-                        raise Error(msg)
-            # Check out the branch/tag/ref we're building; we're looking for the
-            # spec file in the working dir, not the archive. Can't check it out
-            # directly in git clone (with "--branch") b/c on el6 git versions
-            # that doesn't work on non-branches (e.g. tags).
-            rc = utils.unchecked_call(["git", "checkout", "-q", git_hash])
-            if rc:
-                raise Error("Unable to check out %s for some reason." % git_hash)
-
-            full_dest_file = os.path.join(destdir, dest_file)
-            prefix = dest_file[:-4]  # the tar file minus the .tar
-            rc = utils.unchecked_call(["git", "archive", "--format=tar", "--prefix=%s/" % prefix, git_hash, "--output=%s" % full_dest_file])
-            if rc:
-                raise Error("Failed to create an archive of hash %s" % git_hash)
-
-            # git-archive can't create .tar.gz files on git 1.7.1 (SLF 6) so gzip the tar file ourselves
-            # gzip -n will keep hashes of gzips of identical tarballs identical (by
-            # omitting timestamp information)
-            rc = utils.unchecked_call(["gzip", "-fn", full_dest_file])
-            if rc:
-                raise Error("Failed to compress archive at %s" % full_dest_file)
-
-            files = [full_dest_file + ".gz"]
-
-            spec_file = os.path.join(checkout_dir, "rpm", name + ".spec")
-            log.info("Looking for spec file %s in repo", spec_file)
-            if os.path.exists(spec_file):
-                log.info("Found spec file")
-                shutil.copy(spec_file, destdir)
-                files.append(spec_file)
-            else:
-                log.info("Did not find spec file")
-        finally:
-            os.chdir(orig_dir)
+def _mk_prefix(name, tag, tarball):
+    if tarball:
+        if not tarball.endswith('.tar.gz'):
+            raise Error("tarball must end with .tar.gz: '%s'" % tarball)
+        prefix = tarball[:-len('.tar.gz')]
+    else:
+        tag = os.path.basename(tag)
+        tarball_version = re.match(r'(?:v(?=\d))?([^-]+)', tag).group(1)
```

Comment:
I don't understand how this regex works. Can you explain and/or provide an example? In comments.

Reply:
Lies!  But ok